json/src/main/scala/spray/json/JsValue.scala

```diff
@@ -22,6 +22,8 @@ object JsArray {
 
 case object JsNull extends JsValue
 
+case object JsAbsent extends JsValue
```

Comment:
oh, hold on, I just realised what happened here. I thought this code was isolated to the lsp layer. This is quite a big change.

I think the better way to do this would be with a custom data type that has special handling in the `DerivedJsReader`. A more general way to do this would be with scalaz tags.

If we're going down this route, it might be best to just start moving efforts towards creating the scalabus, using Alexy's lsp implementation, and have the bus talk to ensime with JERKY.

I don't really want to start hacking the JSON library ... I plan on moving it out into a separate repository as an example of how to do typeclass derivation with scalaz-deriving and the scope of the library is very much a "simple" json library, whereas circe is appropriate for more complex use cases.

Does what I proposed above make sense?

Reply:
An offtopic sidenote:

> using Alexy's lsp implementation

I see it already second time so I'd like to clarify: if you mean the [lsp4s](https://github.com/scalameta/lsp4s) library, I'm just a mere contributor there, not the primary author.

Reply:
to add some more detail into this advice, I mean adding an ADT like

```scala
sealed abstract class Possibly[A]
final case class There[A](a: A) extends Possibly[A]
final case class Not() extends Possibly[A]
final case class Nullary() extends Possibly[A]
```

and then adding a method to `DerivedJsReader` (and the writer) like

```scala
  implicit def hconsPossibly[A, Key <: Symbol, Value, Remaining <: HList](
    implicit Key: Witness.Aux[Key],
    LV: Lazy[JsReader[Value]],
    DR: DerivedJsReader[A, Remaining]
  ): DerivedJsReader[A, FieldType[Key, Possibly[Value]] :*: Remaining] = { m =>
    val read = ... special handling of the three cases goes here
    field[Key](read) :: DR.readObject(m)
  }
```

I know it's shapeless magic, but it's not so bad :smile:

Reply:
I hear you.  I merely wanted to call attention to the distinction between absent values and null values, and that the current implementation will populate a JsNull for both cases.  I did try to come up with a solution that didn't involve `JsAbsent` but this is the first time I've seen shapeless code.  I'll close the PR without further ado.

Reply:
it's actually a great point! And something I will add to the JSON library when I generalise it. `null` and absent handling is definitely one of the more frustrating things about json. Nothing quite at the level of XML.